### ISSUE TITLE
onClick not handling all values

### ISSUE THREAD

**Issue by totalhack** [formkit/formkit#249]
### Reproduction

https://formkit.link/bafcc2084a9cad6394827c2c9e68deb5

### Describe the bug

I feel like I have to be doing something stupid here. I've been working off the multi-step form example [here](https://github.com/formkit/formkit/issues/63) and sandboxing some new features I need as a proof of concept.  The playground example shows a case where an `onClick` handler doesn't seem to handle being passed an object. Open that example, open the dev console, and click the `LogObj` button.  The definition for that button is:

```javascript
    {
      $formkit: 'button',
      onClick: '$log({"nextStep":"bla", "test":2})',
      children: 'LogObj'
    },
```

The log function is defined in Playground.vue as:

```javascript
  log: target => () => {
    console.log('target:', target)
  },
```

In the console I just see: `target: {"nextStep":"bla"`.

At one point `onClick` was also handling URLs weird for me but now I can't reproduce that in the playground. Is it possible formkit's parsing of variables in schema values is having trouble with certain values/types being used there? Am I just wrong in assuming it can evaluate a JS object? Maybe I need to be passing a JSON string and decoding it if I want to get away with something like the above?

### Environment

• OS: MacOS
• Browser: Chrome 
• Version: 101.0.4951.64 (Official Build) (x86_64)

**Comment by justin-schroeder** [comment]
Schema does not support object notation — it looks a bit like javascript, but it is very much *not javascript* — lots of reasons for this, but one of the biggest is we do not use `eval` for CSP reasons.

You can pass an object to a function, but it needs to be stored in a variable first `$log($someObject)`.

**Comment by totalhack** [comment]
Thanks @justin-schroeder. Do you have any other details, docs, or code you can point me to that lays out the limitations? I'm just thinking ahead to cases where users maybe can not add values to a variable, such as if the form was being completely described and designed from the schema. Does anything in the parsing prevent passing a JSON string representing an object, for example? That seemed to work with the one example I tried.

**Comment by justin-schroeder** [comment]
Yes! The docs on this are pretty comprehensive, and haha, to be clear by comprehensive I mean if its not in the docs, it doesn't do it — not that we have perfect or complete docs 😂. As far as schema "logic" goes, there are basically 4 things you can do:

References: https://formkit.com/advanced/schema#references
Expressions: https://formkit.com/advanced/schema#expressions
Conditionals: https://formkit.com/advanced/schema#conditionals
Loops: https://formkit.com/advanced/schema#loops

References allow you to references data and call functions. Expressions let you perform boolean logic operators, and arithmetic. Conditionals let you make schemas dynamically render some blocks and not others. Loops let you write loops. This combination of items allows nearly Turing complete logic in the schema, but doesn't get into the JavaScript weeds.

The list of things you cannot do in schema that you can do in native JavaScript is pretty much unending, so not worth enumerating.

Nothing prevents passing string literals, those strings could be stringified JSON if you want, but, like with any language you'll need to be careful about all the quotes and escapes and what not. Either way, I would highly recommend having a "data" and a "schema" object in logical implementation of the schema. The best way to think of the schema is as a vue `<template>`. Vue templates are not dynamic at all, they are compiled down to render functions that cannot be modified after the fact, but you can interact with that template by providing it data and methods, and then using things like `v-if` and `v-for` in the template to interact with that data. Exact same story with a schema + data. Think of the schema as a static template that you provide data and functions to and you'll have a much better time with it.

**Comment by totalhack** [comment]
Thanks @justin-schroeder. I'm mostly concerned with the use case of doing *everything* from the schema, because I want non-engineers to be able to edit form definitions, whether that's tweaking the JSON directly at the start or eventually some UI on top of that which modifies the JSON definition to save in a database/etc.  

That said, the use case of needing to pass a custom object in a schema definition is probably something I can avoid (or hide once a UI layer is building the JSON). Maybe this is a bad idea and you could tell me why, but perhaps it would be useful to be able to define a separate data section in the schema which gets merged into the `data` that is accessible to the rest of the schema. So I could define something like:

```javascript
const schema = [
  {
    $cmp: 'FormKit',
    props: {
      type: 'form',
      id: 'form',
      onSubmit: '$submit',
      plugins: '$plugins',
      actions: false,
      data: {
         redirectUrl: `https://www.google.com`
      }
    },
```

And then later in the schema `$redirectUrl` could be used. This would be better than having to hardcode any repeated string literals that may get used such as if I had a `redirect` utility called in multiple places/conditions.

**Comment by andrew-boyd** [comment]
Hey @totalhack — separate from any schema modification if you’re just trying to access literal strings such as a hardcoded URL perhaps you could use a group and name it something like “meta” and fill it with type `hidden` fields that contain those values you could access such as `meta.redirectUrl`?

**Comment by totalhack** [comment]
Thanks @andrew-boyd, I'll consider that as a workaround -- wouldn't that mean those hidden fields also get posted with the form though? Can just ignore that in most cases, depends how strict the endpoint is about the format being posted too of course.

**Comment by andrew-boyd** [comment]
That’s correct, but you get a non-reactive copy the form data passed to your submit handler so you could omit the meta group from that object before posting if your endpoint is strict about it.

**Comment by totalhack** [comment]
> Hey @totalhack — separate from any schema modification if you’re just trying to access literal strings such as a hardcoded URL perhaps you could use a group and name it something like “meta” and fill it with type `hidden` fields that contain those values you could access such as `meta.redirectUrl`?

Hey @andrew-boyd, just wanted to note that I took a different approach to this. Instead I just added a "fake" node to the schema that simply holds metadata, and in a wrapper to `FormKitSchema` I look for that and update a `meta` field in the `data` value.  For example:

```javascript
const schema = [
  {
    type: 'meta',
    data: {
      someTestField: "Hey you can access this elsewhere as $meta.someTestField"
    }
  },
```

It doesn't seem that formkit minds that I put an invalid node in the schema, it just ignores it. This is a lot cleaner than adding groups and hidden inputs IMO.
